GenericList/string_list.h

```diff
@@ -1,31 +1,142 @@
+
+
+    //Remove all elements - maintain capacity
+    void Clear();
+
+    //Shrink array capacity down to count
+    void ShrinkToFit();
+
+    int Capacity() const;
+    int Count() const;
+
+    //Sets the capacity of the internal array to new_capacity. If new_capacity is smaller than Count, do nothing.
+    void Capacity(int new_capacity);
+
+    std::string ToString() const;
+
+    //Take in any amount of arguments of any type, then unpack on element construction - allows for creating new data without checking for logic errors (the compiler and the element's constructor will take care of that)
+    template<typename... Args>
+    void Add(Args&&... args) {
+        if (capacity == count) Capacity(std::max(2 * capacity, 1));
+        new (data + count++) std::string(std::forward<Args>(args)...);
+    }
+
+    std::string* Find(const std::string& val);
+    const std::string* Find(const std::string& val) const;
+
+    template <typename Predicate>
+    std::string* FindIf(Predicate&& pred) {
+        for (auto ptr = begin(); ptr < end(); ptr++) {
+            if (pred(*ptr)) return ptr; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
 
-    std::string ToString();
+    template <typename Predicate>
+    const std::string* FindIf(Predicate&& pred) const {
+        for (auto ptr = begin(); ptr < end(); ptr++) {
+            if (pred(*ptr)) return ptr; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
 
-    bool Add(std::string new_val);
-    bool Contains(std::string val);
-    bool RemoveAt(int index);
+    bool Contains(const std::string& val) const;
+    void RemoveAt(int index);
+
+    int IndexOf(const std::string& val) const;
+
+    //Returns true if a relevant element exists, and removes it.
+    size_t Remove(const std::string& val);
+    
+    template <typename Predicate>
+    size_t RemoveIf(Predicate&& pred) {
+        //Dumb version - O(n^2) time, O(1) space (if removal is O(1) space)
+        /*
+         *  std::string* ptr;
+         *  int deletions = 0;
+         *  while ((ptr = FindIf(pred)) != nullptr) {
+         *      RemoveAt(ptr - data);
+         *      deletions++;
+         *  }
+         *  return deletions;
+         */
+        //idea: double pointers, picker points to next element to check,
+        //                       placer points to next available space
+        size_t new_count = 0;
+        auto picker = begin(), placer = begin();
+        for (; picker < end(); picker++) {
+            if (!pred(*picker)) {
+                std::swap(*placer, *picker);
+                placer++;
+                new_count++;
+            }
+        }
+        //destruct afterwards to ensure std::swap operates on instantiated objects
+        for (; placer < end(); placer++) {
+            placer->~basic_string();
+        }
+
+        size_t removed = count - new_count;
+        count = new_count;
+
+        return removed;
```

Comment:
Nit, you don't need the swap, since you're going to destroy the element in `placer`, can just move assign
```suggestion
auto i = FindIf(pred);
if (i == nullptr) {
    return 0;
}

for (auto j = i + 1; j < end(); ++j) {
    if (!pred(*j)) {
        *i = std::move(*j);
        ++i;
    }
}

for (auto k = i; k < end(); ++k) {
    k->~string();
}

const auto removed = end() - i;
count -= removed;

return removed;
```